File: VodArchiver/VodArchiver/UserInfo/TwitchUserInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml;
using VodArchiver.VideoInfo;

namespace VodArchiver.UserInfo {
	public class TwitchUserInfo : IUserInfo {
		public override ServiceVideoCategoryType Type => Highlights ? ServiceVideoCategoryType.TwitchHighlights : ServiceVideoCategoryType.TwitchRecordings;
		public override string UserIdentifier => Username;
		public override bool Persistable { get => _Persistable; set => _Persistable = value; }
		public override bool AutoDownload { get => _AutoDownload; set => _AutoDownload = value; }
		public override DateTime LastRefreshedOn { get => _LastRefreshedOn; set => _LastRefreshedOn = value; }

		private bool _Persistable;
		private bool _AutoDownload;
		private DateTime _LastRefreshedOn = Util.DateTimeFromUnixTime( 0 );

		private string Username;
		private long? UserID;
		private bool Highlights;

		public TwitchUserInfo( string username, bool highlights ) {
			Username = username;
			UserID = null;
			Highlights = highlights;
			_Persistable = false;
			_AutoDownload = false;
		}

		public TwitchUserInfo( XmlNode node ) {
			ServiceVideoCategoryType t = (ServiceVideoCategoryType)Enum.Parse( typeof( ServiceVideoCategoryType ), node.Attributes["_type"].Value );
			if ( t == ServiceVideoCategoryType.TwitchRecordings ) {
				Highlights = false;
			} else if ( t == ServiceVideoCategoryType.TwitchHighlights ) {
				Highlights = true;
			} else {
				throw new Exception( "Invalid type for TwitchUserInfo." );
			}

			_AutoDownload = node.Attributes["autoDownload"].Value == "true";
			UserID = node.Attributes["userID"].Value == "?" ? null : (long?)long.Parse( node.Attributes["userID"].Value );
			_LastRefreshedOn = Util.DateTimeFromUnixTime( ulong.Parse( node.Attributes["lastRefreshedOn"].Value ) );
			Username = node.Attributes["username"].Value;
			_Persistable = true;
		}

		public override XmlNode Serialize( XmlDocument document, XmlNode node ) {
			node.AppendAttribute( document, "_type", Type.ToString() );
			node.AppendAttribute( document, "autoDownload", AutoDownload ? "true" : "false" );
			node.AppendAttribute( document, "userID", UserID == null ? "?" : UserID.ToString() );
			node.AppendAttribute( document, "lastRefreshedOn", LastRefreshedOn.DateTimeToUnixTime().ToString() );
			node.AppendAttribute( document, "username", Username.ToString() );
			return node;
		}

		public override async Task<FetchReturnValue> Fetch( int offset, bool flat ) {
			List<IVideoInfo> videosToAdd = new List<IVideoInfo>();
			bool hasMore = true;
			long maxVideos = -1;
			int currentVideos = -1;

			if ( UserID == null ) {
				UserID = await TwitchV5.GetUserIdFromUsername( Username, Util.TwitchClientId );
			}
			TwitchVodFetchResult broadcasts = await TwitchV5.GetVideos( UserID.Value, Highlights, offset, 25, Util.TwitchClientId );
			hasMore = offset + broadcasts.Videos.Count < broadcasts.TotalVideoCount;
			maxVideos = broadcasts.TotalVideoCount;

			foreach ( var v in broadcasts.Videos ) {
				videosToAdd.Add( new TwitchVideoInfo( v, StreamService.Twitch ) );
				videosToAdd.Add( new TwitchVideoInfo( v, StreamService.TwitchChatReplay ) );
			}
			currentVideos = broadcasts.Videos.Count;

			if ( videosToAdd.Count <= 0 ) {
				return new FetchReturnValue { Success = true, HasMore = false, TotalVideos = maxVideos, VideoCountThisFetch = 0, Videos = videosToAdd };
			}

			return new FetchReturnValue { Success = true, HasMore = hasMore, TotalVideos = maxVideos, VideoCountThisFetch = currentVideos, Videos = videosToAdd };
		}
	}
}

File: native/vodarchiver/userinfo/twitch-user-info.cpp
#include "twitch-user-info.h"

#include <cstdint>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "vodarchiver/twitch_util.h"
#include "vodarchiver/videoinfo/i-video-info.h"
#include "vodarchiver/videoinfo/twitch-video-info.h"

namespace VodArchiver {
TwitchUserInfo::TwitchUserInfo() = default;

TwitchUserInfo::TwitchUserInfo(std::string username, bool highlights)
  : Username(std::move(username)), Highlights(highlights) {}

TwitchUserInfo::~TwitchUserInfo() = default;

ServiceVideoCategoryType TwitchUserInfo::GetType() {
    return Highlights ? ServiceVideoCategoryType::TwitchHighlights
                      : ServiceVideoCategoryType::TwitchRecordings;
}

std::string TwitchUserInfo::GetUserIdentifier() {
    return Username;
}

FetchReturnValue TwitchUserInfo::Fetch(size_t offset, bool flat) {
    std::vector<std::unique_ptr<IVideoInfo>> videosToAdd;
    bool hasMore = true;
    int64_t maxVideos = -1;
    int64_t currentVideos = -1;

    std::string twitchClientId = "";
    std::string twitchClientSecret = "";

    if (!UserID.has_value()) {
        auto uid = Twitch::GetUserIdFromUsername(Username, twitchClientId, twitchClientSecret);
        if (!uid) {
            return FetchReturnValue{.Success = false, .HasMore = false};
        }
        UserID = *uid;
    }

    auto broadcasts =
        Twitch::GetVideos(*UserID, Highlights, offset, 25, twitchClientId, twitchClientSecret);
    if (!broadcasts) {
        return FetchReturnValue{.Success = false, .HasMore = false};
    }
    hasMore = (offset + broadcasts->Videos.size()) < broadcasts->TotalVideoCount;
    maxVideos = broadcasts->TotalVideoCount;

    currentVideos = broadcasts->Videos.size();
    for (auto& v : broadcasts->Videos) {
        auto tv = std::make_unique<TwitchVideoInfo>();
        auto tc = std::make_unique<TwitchVideoInfo>();
        tv->_Service = StreamService::Twitch;
        tc->_Service = StreamService::TwitchChatReplay;
        tv->_Video = v;
        tc->_Video = std::move(v);
        videosToAdd.push_back(std::move(tv));
        videosToAdd.push_back(std::move(tc));
    }

    if (videosToAdd.size() <= 0) {
        return FetchReturnValue{.Success = true,
                                .HasMore = false,
                                .TotalVideos = maxVideos,
                                .VideoCountThisFetch = 0,
                                .Videos = std::move(videosToAdd)};
    }

    return FetchReturnValue{.Success = true,
                            .HasMore = hasMore,
                            .TotalVideos = maxVideos,
                            .VideoCountThisFetch = currentVideos,
                            .Videos = std::move(videosToAdd)};
}
} // namespace VodArchiver
